src/ngraph_backend_manager.h

```diff
@@ -0,0 +1,65 @@
+/*******************************************************************************
+ * Copyright 2018 Intel Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *******************************************************************************/
+
+#pragma once
+
+#include <memory>
+#include <mutex>
+#include <string>
+#include <tuple>
+
+#include "ngraph/runtime/backend.hpp"
+
+namespace tensorflow {
+namespace ngraph_bridge {
+
+// Manages the nGraph bridge's notion of the current execution backend.
+class BackendManager {
+ public:
+  // Returns the global BackendManager instance.
+  static BackendManager* Instance();
+
+  // Sets the manager's current backend.
+  //
+  // Throws an exception if the supplied configuration string can't be
+  // resolved to an nGraph backend -- either if it doesn't match an
+  // available backend, or if the loaded backend is unable to process
+  // the backend-specific configuration component.
+  //
+  // If the configuration cannot be resolved, the manager's current
+  // backend will remain unchanged.
+  void SetBackendConfig(const std::string& config);
+
+  // Gets the manager's current backend.
+  //
+  // If the backend has not previously been loaded, the implementation
+  // will attempt to load it, using the configuration specified in the
+  // NGRAPH_TF_BACKEND environment variable, or using "CPU" as a
+  // fallback if NGRAPH_TF_BACKEND is not set.
+  //
+  // The second member of the tuple will be true iff the manager is
+  // using the CPU backend.
+  std::tuple<std::shared_ptr<ngraph::runtime::Backend>, bool> GetBackend();
```

Comment:
`SetBackendConfig` has a nice clean interface but `GetBackend` is slightly overloaded, since it attempts to also load a backend if nothing is present. Perhaps a more descriptive name for what it is doing now is`SetBackendFromEnvironment`. In which case to make things consistent, can we have `SetBackendConfig` also return a tuple that `GetBackend` is currently returning

 ideally perhaps: a pure setter (which tries to load from environment if config is not specified) + a pure getter (which throws errors if nothing is set) + pure boolean `isCPUBackend(bool* b)`.

Reply:
I think the caller needs/wants a function somewhere whose functionality is, "Get me the right nGraph backend to use, and I don't care what you have to do in order to get it; I don't care whether you already have one, or if you've been given a config string, or if you're going to check an environment variable, or pull in something from dconf, or check the Registry, or whatever you might want to do on MacOS -- just get me something I can use, and throw an exception if you can't get it.  Oh, and I need to know whether it's the CPU backend."

So that's what `GetBackend()` is doing (and why it's not `backend() const`.)

I did think about splitting this up: we'd have a `BackendState` object, with a Meyers singleton instance accessor.  The state object would just be a struct with the mutex, config string, and current backend.  We could then have free functions that would manipulate the state object, basically exactly as they are now.

The downside is that it depends on all callers getting the manipulations correct.  It seems more obviously-correct to me to specialize the state object a bit for the current callers, keeping the internals private and exposing a couple of accessor APIs that are easy to validate.  But I agree, the resulting class is very specialized for its current role in nGraph-TF; it's not very OO, it's not very reusable, it's just a little bit of glue logic for this one particular situation.

Overall... I'm pretty comfortable either way.  My on-the-fly "which way should we go" landed on the specialized object implementation, but if you'd prefer a more OO approach, I'm happy to write it up that way, too.

(FWIW -- things calling `SetBackendConfig` really don't care about actually getting the current backend.)

Reply:
I do see your point of `Get me the right nGraph backend to use, and I don't care what you have to do in order to get it`. I am good either way. I'll let @avijit-nervana comment.

Reply:
In my opinion, it's better to keep functions simple and do what the name suggests. If the purpose of the function is to get a backend as requested by the caller - then that's what it needs to return. So - need to add the config parameter here and then we can have another method called GetDefaultBackend() or GetCpuBackend().

Reply:
But then we'd need some place to store that config, and then _that_ becomes the global object... :-)

Okay -- I'm thinking it will be more clear what's going on if I break apart the data storage and the manipulation logic.  I'm in the middle of some other changes, but should have an update to this within the next day or two.